File: backend/app/schemas/leave.py
from datetime import datetime
from pydantic import BaseModel


class LeaveRequestCreate(BaseModel):
    leave_type: str  # personal/sick/annual/comp/other
    start_date: str  # "2026-03-28"
    end_date: str
    days: float
    reason: str


class LeaveApproveRequest(BaseModel):
    action: str  # "approve" or "reject"
    remark: str = ""


class LeaveRequestResponse(BaseModel):
    id: str
    user_id: str
    user_name: str | None = None
    leave_type: str
    leave_type_label: str = ""
    start_date: str
    end_date: str
    days: float
    reason: str
    status: str
    status_label: str = ""
    approver_name: str | None
    approve_remark: str | None
    approved_at: datetime | None
    created_at: datetime

    class Config:
        from_attributes = True


class LeaveListResponse(BaseModel):
    total: int
    items: list[LeaveRequestResponse]


class LeaveBalanceResponse(BaseModel):
    leave_type: str
    leave_type_label: str
    total_days: float
    used_days: float
    remaining_days: float

File: web/src/api/leave.ts
import client from './client';

export interface LeaveRequest {
  id: string;
  user_id: string;
  user_name: string | null;
  leave_type: string;
  leave_type_label: string;
  start_date: string;
  end_date: string;
  days: number;
  reason: string;
  status: string;
  status_label: string;
  approver_name: string | null;
  approve_remark: string | null;
  approved_at: string | null;
  created_at: string;
}

export interface LeaveListResponse {
  total: number;
  items: LeaveRequest[];
}

export async function getLeaveRequests(params: {
  page?: number;
  page_size?: number;
  status?: string;
  user_id?: string;
}): Promise<LeaveListResponse> {
  const res = await client.get<LeaveListResponse>('/leave/requests', { params });
  return res.data;
}

export async function approveLeave(id: string, action: 'approve' | 'reject', remark: string = '') {
  const res = await client.post(`/leave/${id}/approve`, { action, remark });
  return res.data;
}

export async function setLeaveBalance(userId: string, year: number, leaveType: string, totalDays: number) {
  const res = await client.put(`/leave/balance/${userId}`, null, {
    params: { year, leave_type: leaveType, total_days: totalDays },
  });
  return res.data;
}
